_posts/2020-08-26-springboot-docker-image.md

```diff
@@ -0,0 +1,399 @@
+ARG JAR_FILE=target/*.jar
+COPY ${JAR_FILE} application.jar
+RUN java -Djarmode=layertools -jar application.jar extract
+
+# the second stage of our build will copy the extracted layers
+FROM adoptopenjdk:14-jre-hotspot
+WORKDIR application
+COPY --from=builder application/dependencies/ ./
+COPY --from=builder application/spring-boot-loader/ ./
+COPY --from=builder application/snapshot-dependencies/ ./
+COPY --from=builder application/application/ ./
+ENTRYPOINT ["java", "org.springframework.boot.loader.JarLauncher"]
+
+```
+We save this configuration in a separate file - `Dockerfile2`.
+
+We build the Docker Image using the command:
+```
+docker build -f Dockerfile2 -t usersignup:v1 .
+```
+After running this command, we get this output: 
+```shell
+Sending build context to Docker daemon  20.41MB
+Step 1/12 : FROM adoptopenjdk:14-jre-hotspot as builder
+14-jre-hotspot: Pulling from library/adoptopenjdk
+.
+.
+Successfully built a9ebf6970841
+Successfully tagged userssignup:v1
+```
+We can see the Docker Image is created with an Image ID and then tagged. 
+
+We finally run the Dive command as before to check the layers inside the generated Docker Image. We can specify either the Image ID or tag as input to the Dive command:
+```shell
+dive userssignup:v1
+```
+As we can see in the output, the layer containing the application is only 11 kB now with the dependencies cached in separate layers.
+![dive screenshot](/assets/img/posts/springboot-docker-image/dive2.png)
+
+## Extracting Internal Dependencies in Separate Layers
+
+We can further reduce the application layer size by extracting any of our custom dependencies in a separate layer instead of packaging them with the application by declaring them in a `yml` like file named `layers.idx`:
+
+```layers.idx
+- "dependencies":
+  - "BOOT-INF/lib/"
+- "spring-boot-loader":
+  - "org/"
+- "snapshot-dependencies":
+- "custom-dependencies":
+  - "io/myorg/"
+- "application":
+  - "BOOT-INF/classes/"
+  - "BOOT-INF/classpath.idx"
+  - "BOOT-INF/layers.idx"
+  - "META-INF/"
+
+```
+In this file -`layers.idx` we have added a custom dependency with the name `io.myorg` containing organization dependencies pulled from a shared repository.
+
+## Conclusion
+
+In this article, we looked at using Cloud-Native Buildpacks to create the container image directly from source code. This is an alternative to building the container image using the conventional way, by first building the fat executable jar and then package it in a container image by specifying the instructions in a Dockerfile. 
+
+We also looked at optimizing our container at runtime by extracting the dependencies in separate layers that get cached in the host and the thin layer of application is downloaded during scheduling in container runtime engines. 
+
+We built the final image in two stages using a multi-stage build. The first stage extracts the dependencies and the second stage copies the extracted dependencies to the final image.
```

Comment:
As mentioned in the previous comment, I don't see the motivation for this. I would delete it, or do you have an idea to motivate this better in the article?